cpp/src/arrow/csv/reader.cc

```diff
@@ -0,0 +1,399 @@
+      }
+    }
+
+    // Finish conversion, create schema and table
+    RETURN_NOT_OK(task_group_->Finish());
+    return MakeTable(out);
+  }
+};
+
+/////////////////////////////////////////////////////////////////////////
+// Parallel TableReader implementation
+
+class ThreadedTableReader : public BaseTableReader {
+ public:
+  ThreadedTableReader(MemoryPool* pool, std::shared_ptr<io::InputStream> input,
+                      ThreadPool* thread_pool, const ReadOptions& read_options,
+                      const ParseOptions& parse_options,
+                      const ConvertOptions& convert_options)
+      : BaseTableReader(pool, read_options, parse_options, convert_options),
+        thread_pool_(thread_pool) {
+    // Readahead one block per worker thread
+    int32_t block_queue_size = thread_pool->GetCapacity();
+    readahead_ = std::make_shared<ReadaheadSpooler>(
+        pool_, input, read_options_.block_size, block_queue_size, kDefaultLeftPadding,
+        kDefaultRightPadding);
+  }
+
+  ~ThreadedTableReader() {
+    if (task_group_) {
+      // In case of error, make sure all pending tasks are finished before
+      // we start destroying BaseTableReader members
+      ARROW_UNUSED(task_group_->Finish());
+    }
+  }
+
+  Status Read(std::shared_ptr<Table>* out) {
+    task_group_ = internal::TaskGroup::MakeThreaded(thread_pool_);
+    Chunker chunker(parse_options_, static_cast<int32_t>(read_options_.num_rows));
+
+    // Get first block and process header serially
+    RETURN_NOT_OK(ReadNextBlock());
+    if (eof_) {
+      return Status::Invalid("Empty CSV file");
+    }
+    RETURN_NOT_OK(ProcessHeader());
+
+    while (!eof_ && task_group_->ok()) {
+      // Consume current chunk
+      uint32_t chunk_size = 0;
+      RETURN_NOT_OK(chunker.Process(reinterpret_cast<const char*>(cur_data_),
+                                    static_cast<uint32_t>(cur_size_), &chunk_size));
+      int32_t chunk_rows = chunker.num_rows();
+      if (chunk_rows > 0) {
+        // Got a chunk of rows
+        const uint8_t* chunk_data = cur_data_;
+        std::shared_ptr<Buffer> chunk_buffer = cur_block_;
+        int64_t chunk_index = cur_block_index_;
+
+        // "mutable" allows to modify captured by-copy chunk_buffer
+        task_group_->Append([=]() mutable -> Status {
+          auto parser =
+              std::make_shared<BlockParser>(parse_options_, num_cols_, chunk_rows);
+          uint32_t parsed_size = 0;
+          RETURN_NOT_OK(parser->Parse(reinterpret_cast<const char*>(chunk_data),
+                                      chunk_size, &parsed_size));
+          if (parsed_size != chunk_size || chunk_rows != parser->num_rows()) {
+            return Status::Invalid("Chunker and parser disagree on block size");
+          }
+          RETURN_NOT_OK(ProcessData(parser, chunk_index));
+          // Keep chunk buffer alive within closure and release it at the end
+          chunk_buffer.reset();
```

Comment:
May want to create some kind of a BufferPool abstraction so that you can return a buffer to the common pool rather than deallocating, then it would make this bit a bit cleaner perhaps